lib/Service/MailSearch.php

```diff
@@ -0,0 +1,152 @@
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\ILogger;
+use function array_keys;
+use function array_reverse;
+use function in_array;
+use function uasort;
+
+class MailSearch implements IMailSearch {
+
+	/** @var IMAPClientFactory */
+	private $clientFactory;
+
+	/** @var SearchStrategyFactory */
+	private $searchStrategyFactory;
+
+	/** @var SearchFilterStringParser */
+	private $filterStringParser;
+
+	/** @var MailboxMapper */
+	private $mailboxMapper;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(IMAPClientFactory $clientFactory,
+								SearchStrategyFactory $searchStrategyFactory,
+								SearchFilterStringParser $filterStringParser,
+								MailboxMapper $mailboxMapper,
+								ILogger $logger) {
+		$this->clientFactory = $clientFactory;
+		$this->searchStrategyFactory = $searchStrategyFactory;
+		$this->filterStringParser = $filterStringParser;
+		$this->mailboxMapper = $mailboxMapper;
+		$this->logger = $logger;
+	}
+
+	/**
+	 * @param Account $account
+	 * @param string $mailboxName
+	 * @param string|null $filter
+	 * @param string|null $cursor
+	 *
+	 * @return IMAPMessage[]
+	 * @throws ServiceException
+	 */
+	public function findMessages(Account $account, string $mailboxName, ?string $filter, ?int $cursor): array {
+		$client = $this->clientFactory->getClient($account);
+		try {
+			$mailbox = $this->mailboxMapper->find($account, $mailboxName);
+		} catch (DoesNotExistException $e) {
+			throw new ServiceException('Mailbox does not exist', 0, $e);
+		}
+
+		try {
+			$query = $this->filterStringParser->parse($filter);
+
+			// In flagged we don't want anything but flagged messages
+			if ($mailbox->isSpecialUse(Horde_Imap_Client::SPECIALUSE_FLAGGED)) {
+				$query->flag(Horde_Imap_Client::FLAG_FLAGGED);
+			}
+
+			// Don't show deleted messages unless for folders
+			if (!$mailbox->isSpecialUse(Horde_Imap_Client::SPECIALUSE_TRASH)) {
+				$query->flag(Horde_Imap_Client::FLAG_DELETED, false);
+			}
+
+			$ids = $this->searchStrategyFactory
+				->getStrategy($client, $mailbox->getMailbox(), $query, $cursor)
+				->getIds(20);
+		} catch (Horde_Imap_Client_Exception $e) {
+			throw new ServiceException('Could not get message IDs: ' . $e->getMessage(), 0, $e);
+		}
+
+		try {
+			$fetchQuery = new Horde_Imap_Client_Fetch_Query();
+			$fetchQuery->envelope();
+			$fetchQuery->flags();
+			$fetchQuery->size();
+			$fetchQuery->uid();
+			$fetchQuery->imapDate();
+			$fetchQuery->structure();
+			$fetchQuery->headers(
+				'imp',
+				[
+					'importance',
+					'list-post',
+					'x-priority',
+					'content-type',
+				],
+				[
+					'cache' => true,
+					'peek' => true
+				]
+			);
+
+			$fetchResult = $client->fetch($mailbox->getMailbox(), $fetchQuery, ['ids' => $ids]);
+		} catch (Horde_Imap_Client_Exception $e) {
+			throw new ServiceException('Could not fetch messages', 0, $e);
+		}
+
+		// TODO: do we still need this fix?
+		ob_start(); // fix for Horde warnings
```

Comment:
Urgh :see_no_evil: Do you know some of these warnings?

Reply:
NONE. That is why I added the `todo` ;)

Reply:
It's from https://github.com/nextcloud/mail/commit/6baf36b844ddd7dfa414515c73bfca3063249631. This could be totally useless by know.